### ISSUE TITLE
`breakanywhere` is incompatible with UTF-8 

### ISSUE THREAD

**Issue by ScumCoder** [gpoore/minted#123]
The title says it all. If there are Unicode symbols in the source that needs to be printed with `minted`, everything works until `breakanywhere` is used.

SSCCEs:
[No unicode, `breakanywhere` is on.](https://gist.github.com/ScumCoder/03cba04492e73ac544155dc03bf991de/5e1d9fc1ecc1afa7dbf36362ff617191d8b87cac) Everything is fine.
[Text has unicode, `breakanywhere` is off.](https://gist.github.com/ScumCoder/03cba04492e73ac544155dc03bf991de/ba96677c477e0a318d387033716c6c5787e8e811) File compiles, but the long line is ugly.
[Text has unicode, `breakanywhere` is on.](https://gist.github.com/ScumCoder/03cba04492e73ac544155dc03bf991de/560f71ae05d6e797762c5cf378d8160b8d5b5614) Compilation fails with a load of this:

```
Package inputenc Error: Unicode char �\FancyVerbBreakAnywhereBreak (U+AF9)(inputenc) not set up for use with LaTeX. \PYG{c+c1}{// Привет}
Package inputenc Error: Keyboard character used is undefined(inputenc) in inputencoding `utf8'. \PYG{c+c1}{// Привет}
```
I tried `latex` and `pdflatex` with similar results. Unfortunately I can't use `xelatex` for certain reasons.
Thanks in advance. Any help will be extremely appreciated.

**Comment by gpoore** [comment]
This is due to the way the pdfTeX engine implements UTF-8. I have some ideas about possible fixes, but any solution will be somewhat technical, so it may take some time. LuaTeX should work if it is an option.

For now, you might consider trying `breakafter` or `breakbefore` if you can define a list of ASCII characters that are suitable break locations.

**Comment by ScumCoder** [comment]
[Here](http://stackoverflow.com/questions/2132933) it says that `fancyvrb` (which AFAIU is used by `minted`) does not work well with Unicode.

[One proposed solution](http://www.tug.org/pipermail/pdftex/2003-February/003510.html) goes all the way back to 2003; I checked the `fancyvrb.sty` which is in my TeXLive2016 and the relevant lines are still the same. So I tried to apply the proposed patch; it didn't help.

Another proposed solution is to use `ucs` + `utf8x`, but it is horribly outdated.

**Comment by davidcarlisle** [comment]
This works although a more complete fix would do 3,4,5 byte sequences as well as two, and probably make the redefinition just local to the environment, but I do it globally here. Sorry about my Russian text, blame google translate:-)

```
\documentclass{article}

\usepackage[utf8]{inputenc}
\usepackage[T2A]{fontenc}
\usepackage[english,main=russian]{babel}
\usepackage{minted}

\makeatletter
\let\xUTFviii@two@octets\UTFviii@two@octets

\def\UTFviii@two@octets#1#2{%
\ifx\FancyVerbBreakAnywhereBreak#2%
\expandafter\xUTFviii@two@octets\expandafter#1%
\else
\xUTFviii@two@octets#1#2%
\fi
}
\makeatother
\begin{document}
	
\newminted{cpp}{
	breaklines,
	breakanywhere
}

\begin{cppcode}
	int main(void)
	{
		// Hello
		// длиннаялиниянетместадлиннаялиниянетместадлиннаялиниянетместадлиннаялиниянетместадлиннаялиниянетместадлиннаялиниянетместадлиннаялиниянетместадлиннаялиниянетместадлиннаялиниянетместадлиннаялиниянетместадлиннаялиниянетместа
	}	
\end{cppcode}

\end{document}
```

**Comment by gpoore** [comment]
@ScumCoder The issues you referenced are with `VerbatimOut`. `minted` has actually patched `VerbatimOut` internally with more or less equivalent fixes in every version since 1.7. The issue here is specifically with `breakanywhere`.

The problem is that `breakanywhere` is inserting potential break locations between all tokens (skipping math and groups). If I understand things correctly, XeTeX and LuaTeX are Unicode-aware, so a token corresponds to a Unicode code point. But pdfTeX works at the level of raw bytes, so a token corresponds to a byte, and then `breakanywhere` breaks up multi-byte code points by inserting potential breaks within them.

@davidcarlisle Thanks for that. I was thinking about the painful possibility of adding multi-byte lookahead to the `breakanywhere` scanner, but locally redefining the UTF-8 macros would be a lot easier.

Just for future reference, the fix for this will ultimately go into my new [`fvextra` package](https://github.com/gpoore/fvextra), which is a collection of extensions and patches for `fancyvrb`. I'm pulling various `fancyvrb` extensions and patches out of `minted` and `pythontex`, and putting them in their own package to simplify things. `fvextra` can already be loaded before `minted` to provide the few new features that `minted` lacks, and the next version of `minted` will require `fvextra`.

**Comment by davidcarlisle** [comment]
On 29 Ju
>
>
> @davidcarlisle <https://github.com/davidcarlisle> Thanks for that. I was
> thinking about the painful possibility of adding multi-byte lookahead to
> the breakanywhere scanner, but locally redefining the UTF-8 macros would
> be a lot easier.
>

 actually that is probably conceptually better I started to do it but it
was easier for me to hack the utf8 macros (since I wrote them:-)  it
shouldn't be too painful: you wouldn't have to teach it much about utf8
just that if the token you have got is an active character with definition
starting with `\UTFviii@two@octets` then it's definition will be

`\UTFviii@two@octets <same character>`

in which case you need to grab one more token to keep them together, and
similarly if it is

`\UTFviii@three@octets <same character>`

you need to grab the next two tokens  so as to keep a three byte utf8 code
together, prefixed by `\UTFviii@three@octets`

but letting the scanner put in the bad tokens then filtering them out also
seems to work, so whichever is more robust:-)


David

Just for future reference, the fix for this will ultimately go into my new
> fvextra package <https://github.com/gpoore/fvextra>, which is a
> collection of extensions and patches for fancyvrb. I'm pulling various
> fancyvrb extensions and patches out of minted and pythontex, and putting
> them in their own package to simplify things. fvextra can already be
> loaded before minted to provide the few new features that minted lacks,
> and the next version of minted will require fvextra.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gpoore/minted/issues/123#issuecomment-229394487>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ABNcAoIQG-i3NiAAR1kcInXvgoFKv3Tsks5qQpA2gaJpZM4JAoDC>
> .
>



-- 
http://dpcarlisle.blogspot.com/

**Comment by ScumCoder** [comment]
@davidcarlisle thanks a bunch.

**Comment by ScumCoder** [comment]
@davidcarlisle thanks a bunch.

P.S. sorry for closing, I was curious what the "close and comment" button does

**Comment by gpoore** [comment]
The [latest commit for `fvextra`](https://github.com/gpoore/fvextra/issues/1) should fix this. If you download the `fvextra.sty` and put it in your document directory, and `\usepackage{fvextra}` before `\usepackage{minted}`, everything should work. `breakbefore` and `breakafter` were also broken, as might be expected, so I fixed them as well.

The next version of `fvextra` should be on CTAN in the next week or so. And the next version of `minted`, which will require `fvextra`, will follow shortly.

**Comment by gpoore** [comment]
This is fixed in [v2.3](https://github.com/gpoore/minted/releases/tag/v2.3). `minted` v2.3 and `fvextra` v1.1, which it requires, have both been submitted to CTAN.
